man/go-gma-server.6
't
'\" <<ital-is-var>>
'\" <<bold-is-fixed>>
.TH GO-GMA-SERVER 6 "GMA Toolkit 4.2.2" 16-Nov-2020 "Games" \" @@mp@@
.SH NAME
go-gma-server \- GMA battle grid map server
.SH SYNOPSIS
'\" <<usage>>
.na
.B go\-gma\-server
.RB [ \-h ]
.ad
.LP
.na
.B go\-gma\-server
.RB [ \-help ]
.ad
.LP
.na
.B go\-gma\-server
.RB [ \-init\-file
.IR path ]
.RB [ \-log\-file
.IR path ]
'\" .RB [ \-mysql
'\" .IR database ]
.RB [ \-password\-file
.IR pass-file ]
.RB [ \-endpoint
.RI [ host ]\fB:\fP port ]
.RB [ \-save\-interval
.IR interval ]
.RB [ \-sqlite
.IR path ]
.ad
'\" <</usage>>
.SH DESCRIPTION
.LP
The individual
.BR mapper (6)
clients used by players in a game may keep in contact with one another so that they all
display the same contents. A change made on one client (moving a creature token or
adding a room, say) appears on all the others. This is accomplished by starting a
.B go\-gma\-server
process and having all of the 
.B mapper 
clients connect to it via their 
.B \-\-host
and 
.B \-\-port 
options.
.LP
Once connected, the
server
will send an initial greeting that may define a list of player character tokens
to appear on the 
.B mapper
context menus, or any other useful information the clients need to have
at startup time. It may, at the GM's option, even initialize the client to
show the full current game state.
.LP
From that point forward, the server
is a simple echo relay between the clients, so they communicate with each
other via the service. The server also tracks the commands it sees, so that it maintains
a notion of the current state of the game. Clients may re-sync with the server in case
they restart or otherwise miss any updates so they match the server's state.  The server
may respond directly to some client queries
(e.g., 
.RB \*(lq AI? \*(rq)
if it knows the answer rather than referring the query to the other clients.
.LP
To guard against nuisance or malicious port scans and other superfluous connections
which don't proceed with the normal activity, the server will automatically drop
any clients which don't authenticate within a few poll intervals. (In actual production
use, we have observed some automated agents which connected and then sat idle for hours,
consuming server resources.)
.LP
This is a re-implementation from scratch of the GMA server in the Go programming language.
The original Python implementation is documented in
.BR mapservice (6).
.SH OPTIONS
.LP
The following options control the behavior of
.BR go\-gma\-server .
'\" <<list>>
.TP
.BR \-h , \-help
Print a usage summary and exit.
.TP
.BI "\-init\-file " init-file
Each command in
.I init-file
will be sent to each client upon connection to the server.
The lines of this file must therefore be valid mapper protocol commands as documented in
.BR mapper (6).
The JSON data portion of each command may span multiple lines. In
this case, all lines after the first must be indented. The final closing brace
may appear unindented on a line by itself. For example, the line
.RS
.RS
.LP
.na
.nf
.B "\fCAC {\[dq]Name\[dq]:\[dq]Fred\[dq],\[dq]ObjID\[dq]:\[dq]PC123\[dq],\[dq]Color\[dq]:\[dq]red\[dq]}\fR"
.fi
.ad
.RE
.LP
could also be written in the file as:
.na
.nf
.RS
.LP
.ft C
AC {
\ \ \ \[dq]Name\[dq]: \[dq]Fred\[dq],
\ \ \ \[dq]ObjID\[dq]: \[dq]PC123\[dq],
\ \ \ \[dq]Color\[dq]: \[dq]red\[dq]
}
.ft
.fi
.ad
.RE
.LP
Commands which are only sent by clients to the server may not appear in this file.
Note that the protocol recommends that only comments should be sent before authentication,
and even so only
comments,
.BR AC ,
.BR DSM ,
.BR UPDATES ,
and
.B WORLD
commands are allowed to be sent before the server's
.B READY
command.
.LP
The original version of the server only read this file at start-up, repeating its
contents from memory to all clients; this version reads the file every time a client
connects, allowing changes to be made to the initialization file which take effect
immediately.
.LP
The following special commands may appear in this file. These are not sent to
the clients directly, but trigger other setup events in the server itself:
'\" <<desc>>
.TP 15
.B AUTH
If this command appears in the initialization file, only the contents up to
that line are sent before a client successfully authenticates to the server.
After authentication is complete, the remainder of the initialization file
will be sent to the client. 
There is no JSON data following this command.
.TP
.B READY
If this command appears in the initialization file, this indicates that the server's
.B READY
command should be sent to the client at this point. This allows the initialization file
to specify arbitrary server commands automatically beyond the limited set allowed before
the 
.B READY
command is sent. If this does not appear in the file, the server will automatically send
.B READY
when the client is fully signed-on and it is ready to communicate with the client normally.
There is no JSON data following this command.
.TP
.B SYNC
Enables sync-on-connect mode. With this set, the server will automatically perform the
effects of a 
.B SYNC
command for each client as it connects (after authentication if that is required).
There is no JSON data following this command.
.LP
The server commands
.BR CONN ,
.BR DENIED ,
.BR GRANTED ,
.BR OK ,
.BR PRIV ,
and
.B PROTOCOL
may never appear in this file.
.RE
'\" <</>>
.TP
.BI "\-log\-file " log-file
Append a record of server actions and diagnostic messages to the specified file.
By default, this is sent to the standard error device, as is the case if
.I log-file
is explicitly given the value 
.RB \*(lq \- \*(rq
(single hyphen). If
.I log-file
is the empty string, no log information is produced at all.
.RS
.LP
The
.I log-file
string may include a number of special tokens which are replaced with dynamic
values at runtime. These include:
.TS
center;
cfC l cfC l.
%A	full weekday	%p	AM or PM
%a	abbreviated weekday	%R	=\fC%H:%M\fR
%B	full month	%r	=\fC%I:%M:%S %p\fR
%b	abbreviated month	%S	second as 00\-60
%C	year 00\-99	%s	UNIX timestamp as integer
%c	time and date	%T	=\fC%H:%M:%S\fR
%d	date as 01\-31	%U	week as 00\-53 (Sun start)
%e	date as (sp)1\-31	%u	weekday 1=Mon, 7=Sun
%F	=\fC%Y-%m-%d\fR	%V	week as 00\-53 (Mon start)
%H	hour 00\-23	%v	=\fC%e-%b-%Y\fR
%h	=\fC%b\fR	%W	week as 00\-53 (Mon start)
%I	hour as 01\-12	%w	weekday 0=Sun, 6=Sat
%j	day of year as 001\-366	%X	time
%k	hour as (sp)0\-23	%Y	full year
%L	milliseconds as 000\-999	%y	two-digit year
%l	hour as (sp)1\-12	%Z	time zone name
%M	minute as 00\-59	%z	time zone offset
%m	month as 01\-12	%\[mc]	microseconds as 000\-999
%P	process ID	%%	literal \fC%\fR character
.TE
.RE
.TP
.BI "\-password\-file " pass-file
If this option is given, the server will require clients to authenticate with a
valid password. The first line of
.I pass-file
is the shared password expected to be used by all players. If a second line
is present, it provides a GM password which will grant GM privileges to a user
authenticating using it.
.RS
.LP
Subsequent lines, if any, give specific passwords for individual users. These have the
form
.RS
.IB username : password
.RE
Any login for the given 
.I username
must use the individual password for that user to successfully authenticate to the server.
Any
.I username
not listed in the password file (other than
.RB \*(lq gm \*(rq)
will successfully authenticate if the general-use shared password (the first line of
the password file) is given.
.RE
.TP
.BI "\-endpoint " \fR[\fPhost\fR]\fP\fB:\fP port
The service will accept incoming connections on the specified TCP port. The default is \*(lq:2323\*(rq.
.TP
.BI "\-save\-interval " interval
If the
'\" .B \-mysql
'\" or
.B \-sqlite
option is given, this sets the frequency at which the game state is saved
to the database.
The
.I interval
value is a number followed by a unit, or a number of these together in one string
(units are
.RB \*(lq h \*(rq
for hours,
.RB \*(lq m \*(rq
for minutes,
.RB \*(lq s \*(rq
for seconds,
.RB \*(lq ms \*(rq
for milliseconds,
.RB \*(lq us \*(rq
or
.RB \*(lq \[mc]s \*(rq
for microseconds,
and
.RB \*(lq ns \*(rq
for nanoseconds)
such as 
.RB \*(lq 10m30s \*(rq.
The default is 
.RB \*(lq 10m \*(rq
meaning every 10 minutes. 
.TP
.BI "\-sqlite " path
Open the sqlite3 database 
.I path
as the persistent record of game state and history. Any saved game state is read in from
this database when the service starts, allowing the game to continue from where it was
when the server was stopped. The service will periodically save its current state to this
database.
'\" .TP
'\" .BI "\-mysql " database
'\" Analogous to the
'\" .B \-sqlite
'\" option, this uses a MySQL database for persistent storage. 
'\" <</ital-is-var>>
'\" .I "This is not currently implemented."
'\" <<ital-is-var>>
'\" <</>>
.SH SECURITY
.LP
'\" <</bold-is-fixed>>
.B "The authentication system employed here is simplisitic and not sufficient for general application security,"
but for the sake of inconveniencing cheating at our little game
amongst friends, it will suffice for our narrowly-defined purposes where the stakes
are negligibly low.
'\" <<bold-is-fixed>>
.LP
It is intended just to discourage cheating at the game by looking
at spoilers or direct messages intended for other users, not for any more rigorous
protection.
.LP
The main weakness of the system is that passwords are stored in plaintext on the
server, which means it is critical to secure the password file and the system itself.
Caution your players to use a password for the mapper that is different from any other
passwords they use (which should be the password practice people observe anyway). A
breach that reveals passwords from the server's file would then only allow an imposter
to log in to your map service, which admittedly is more of an inconvenience than a serious security issue, assuming you use your map server just for playing a game and not for
the communication of any sensitive information. 
.LP
Don't use the GMA mapper server for the communication of sensitive information. It's
part of a game. Just play a game with it.
.SH "SIGNALS"
.LP
The map service responds to the following signals while running.
Note that this is different from the behavior of the original
server.
'\" <<desc>>
.TP 8
.B HUP
'\"This signal causes the server to save state if needed and shut down gracefully.
Shuts down the server gracefully.
No new connections will be accepted, but the server will wait for existing ones
to terminate before shutting down.
.TP
.B INT
Emergency shutdown. Just like the graceful shutdown caused by a HUP signal,
but forces all existing connections to immediately terminate.
.TP
.B USR1
Reload password and initialization files. Send this signal after modifying
either of these files when the server is already running.
'\".TP
'\".B USR2
'\"This signal causes the server to save its current state if needed.
'\" <</>>
.SH "SEE ALSO"
.LP
.BR gma (6),
.BR mapper (5),
.BR mapper (6),
.BR mapservice (6).
.SH AUTHOR
.LP
Steve Willoughby / steve@madscience.zone.
.SH BUGS
.SH COPYRGHT
Part of the GMA software suite, copyright \(co 1992\-2021 by Steven L. Willoughby (MadScienceZone), Aloha, Oregon, USA. All Rights Reserved. Distributed under BSD-3-Clause License. \"@m(c)@


CHANGELOG.md
# Changelog
## Current Version Information
 * GMA Core API Library Version: 4.xx.xx	<!-- @@##@@ -->
 * Supported GMA Mapper Version: 4.xx.xx	<!-- @@##@@ -->
 * Supported GMA Mapper Protocol: 400		<!-- @@##@@ -->
 * Supported GMA Mapper File Format: 20		<!-- @@##@@ -->
 * Effective Date: 22-Apr-2022			<!-- @@##@@ -->

## v0.0.0
Added note pointing to paizo's Community Use Policy and GMA's usage of Pathfinder
game-related information.

### Changes
* Switched to server protocol 400. This is a major change, cleaning up the various protocol commands and changing the data marshaling for each command from TCL lists to JSON strings.

## Previous unnumbered versions


cmd/go-gma-server/go-gma-server.go
/*
########################################################################################
#  _______  _______  _______                ___       _______     _______              #
# (  ____ \(       )(  ___  )              /   )     / ___   )   / ___   )             #
# | (    \/| () () || (   ) |             / /) |     \/   )  |   \/   )  |             #
# | |      | || || || (___) |            / (_) (_        /   )       /   )             #
# | | ____ | |(_)| ||  ___  |           (____   _)     _/   /      _/   /              #
# | | \_  )| |   | || (   ) | Game           ) (      /   _/      /   _/               #
# | (___) || )   ( || )   ( | Master's       | |   _ (   (__/\ _ (   (__/\             #
# (_______)|/     \||/     \| Assistant      (_)  (_)\_______/(_)\_______/             #
#                                                                                      #
########################################################################################
*/
//
////////////////////////////////////////////////////////////////////////////////////////
//                                                                                    //
//                                     MapService                                     //
//                                                                                    //
// Inter-map communication service.  Transmits map updates to other maps and allows   //
// API callers to inject events to be sent to all maps.                               //
//                                                                                    //
// This is a re-implementation from scratch of the Python GMA game server (as         //
// originally implemented in the Mapper.MapService module), in the Go language, in    //
// the hopes that this will provide better performance than the Python version.  It   //
// was also done out of personal interest to explore Go design features for a server  //
// such as this one.                                                                  //
//                                                                                    //
////////////////////////////////////////////////////////////////////////////////////////

package main

import (
	"bufio"
	"database/sql"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"

	"internal/mapservice"

	"github.com/MadScienceZone/go-gma-server/mapservice"

	_ "github.com/mattn/go-sqlite3"
)

//
// eventMonitor responds to signals and timers that affect our overall operation
// independent of client requests.
//
func eventMonitor(sig_chan chan os.Signal, stop_chan chan int, ms *mapservice.MapService, saveInterval int) {
	report_interval := 1
	var save_signal *time.Ticker

	if saveInterval <= 0 {
		save_signal = time.NewTicker(1 * time.Minute)
		save_signal.Stop()
		log.Printf("Save interval specified as %d; periodic saves DISABLED", saveInterval)
	} else {
		save_signal = time.NewTicker(time.Duration(saveInterval) * time.Minute)
	}
	ping_signal := time.NewTicker(1 * time.Minute)

	if ms.Database == nil {
		log.Printf("No database open; periodic saves DISABLED")
		save_signal.Stop()
	}

	for {
		select {
		case s := <-sig_chan:
			log.Printf("Received signal %v", s)
			switch s {
			case syscall.SIGHUP:
				stop_chan <- 1

			case syscall.SIGUSR1:
				ms.DumpState()

			case syscall.SIGUSR2:
				log.Printf("**SAVE** due to signal")
				if err := ms.SaveState(); err != nil {
					log.Printf("Error saving game state: %v", err)
				}

			case syscall.SIGINT:
				// Make a quick effort to shut down as fast as possible
				// by terminating all client connections immediately.
				log.Printf("EMERGENCY SHUTDOWN INITIATED")
				ms.AcceptIncoming = false
				for i, client := range ms.Clients {
					log.Printf("Terminating client %v from %s", i, client.ClientAddr)
					client.Connection.Close()
				}
				stop_chan <- 1
			}

		case t := <-save_signal.C:
			// suppress messages and unnecessary saves
			// if we're idling
			if ms.SaveNeeded || report_interval <= 1 {
				log.Printf("***SAVE*** due to timer %v", t)
				if err := ms.SaveState(); err != nil {
					log.Printf("Error saving game state: %v", err)
				}
			}

		case <-ping_signal.C:
			any_connections := ms.PingAll()
			if any_connections {
				if report_interval > 1 {
					report_interval = 1
					log.Printf("Activity detected; reset ping timer to 1 minute")
					ping_signal.Reset(1 * time.Minute)
				}
			} else {
				new_interval := report_interval * 2
				if new_interval > 60 {
					new_interval = 60
				}
				if new_interval != report_interval {
					report_interval = new_interval
					log.Printf("No connections; backing off ping timer to %d minutes", new_interval)
					ping_signal.Reset(time.Duration(new_interval) * time.Minute)
				}
			}
		}
	}
}

func main() {
	var GMAVersionNumber, GMAMapperProtocol string
	var err error

	// Automatically generated version numbers
	GMAVersionNumber = "4.2.2" // @@##@@
	GMAMapperProtocol = "400"  // @@##@@

	passfile := flag.String("password-file", "", "get passwords from the designated file")
	port := flag.Int("port", 2323, "TCP port of map service")
	logfile := flag.String("log-file", "", "log connections and other info to this file")
	initfile := flag.String("init-file", "", "initial commands to send all clients upon authentication")
	greetfile := flag.String("greet-file", "", "initial server greeting sent to clients upon connection")
	saveint := flag.Int("save-interval", 10, "frequency at which to save game state")
	sqlitedb := flag.String("sqlite", "", "use the named sqlite3 database for persistent storage")
	mysqldb := flag.String("mysql", "", "use the named mysql database for persistent storage")
	flag.Parse()

	if *logfile != "" {
		lf, err := os.OpenFile(*logfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			log.Fatalf("Unable to open logfile \"%s\": %v", *logfile, err)
			os.Exit(2)
		}
		log.SetOutput(lf)
		defer lf.Close()
	}

	if GMAMapperProtocol != mapservice.PROTOCOL_VERSION {
		log.Printf("WARNING! This server implements service protocol version %s but %s is the current protocol for the GMA tool suite!\n",
			mapservice.PROTOCOL_VERSION, GMAMapperProtocol)
		expected_proto, err := strconv.Atoi(GMAMapperProtocol)
		supported_proto, err2 := strconv.Atoi(mapservice.PROTOCOL_VERSION)
		if err != nil || err2 != nil {
			log.Fatal("FATAL! Unable to parse version numbers as integers! Bailing out now.")
			os.Exit(1)
		}

		if expected_proto < supported_proto {
			log.Printf("=======  Since our version is higher, this may be an experimental new server.")
		} else {
			log.Printf("=======  This probably means this is an old version of the mapservice program.")
		}
	}

	log.Printf("GMA Map Service (golang), version %s\n", GMAVersionNumber)

	// open database
	var sqldb *sql.DB
	if *sqlitedb != "" {
		if *mysqldb != "" {
			log.Fatalf("You can't specify --sqlite and --mysql at the same time.")
			os.Exit(1)
		}
		if _, err = os.Stat(*sqlitedb); os.IsNotExist(err) {
			// database doesn't exist yet; create a new one

			sqldb, err = sql.Open("sqlite3", "file:"+*sqlitedb)
			log.Printf("No existing sqlite3 database \"%s\" found--creating a new one.", *sqlitedb)
			if err != nil {
				log.Fatalf("Unable to create sqlite3 database %s: %v", *sqlitedb, err)
				os.Exit(2)
			}
			_, err = sqldb.Exec(`
				create table users (
					userid integer primary key,
					username text not null
				);
				create table dicepresets (
					presetid    integer primary key,
					userid      integer not null,
					name        text    not null,
					description text    not null,
					rollspec    text    not null,
						foreign key (userid)
							references users (userid) 
							on delete cascade
				);
				create table events (
					eventid integer primary key,
					rawdata  text    not null,
					sequence integer not null,
					key      text    not null,
					class    text    not null,
					objid    text    not null
				);
				create table extradata (
					extraid integer primary key,
					eventid integer not null,
					datarow text    not null,
						foreign key (eventid)
							references events (eventid)
							on delete cascade
				);
				create table chats (
					rawdata text    not null,
					msgid   text    not null
				);
				create table images (
					name    text    not null,
					zoom    text    not null,
					location text   not null
				);
				create table idbyname (
					name    text    not null,
					objid	text    not null
				);
				create table classbyid (
					objid   text    not null,
					class   text    not null
				);`)
			if err != nil {
				log.Printf("Unable to create sqlite3 database %s contents: %v", *sqlitedb, err)
				log.Fatalf("WARNING! %s may be in a corrupt state--fix or delete before running the server!", *sqlitedb)
				os.Exit(2)
			}
		} else {
			sqldb, err = sql.Open("sqlite3", "file:"+*sqlitedb)
		}
		defer sqldb.Close()
	} else if *mysqldb != "" {
		log.Fatalf("--mysql not yet implemented.")
		os.Exit(1)
	} else {
		log.Printf("WARNING: No database back-end specified. No persistent data storage will be used!")
		log.Printf("(Avoid this by specifying the --sqlite=<filename> option)")
		sqldb = nil
	}

	// set up authentication
	var groupPassword []byte
	var gmPassword []byte
	personalPasswords := make(map[string][]byte)

	if *passfile != "" {
		// The password file contains the group password
		// and (optionally) gm password. Other personal
		// ones are stored in the database (if they are used)
		fp, err := os.Open(*passfile)
		if err != nil {
			log.Fatalf("Unable to open password file \"%s\": %v", *passfile, err)
			os.Exit(2)
		}
		scanner := bufio.NewScanner(fp)
		if scanner.Scan() {
			groupPassword = []byte(scanner.Text())
			if scanner.Scan() {
				gmPassword = []byte(scanner.Text())
				//
				// continue reading personal passwords if any
				//
				line := 3
				for scanner.Scan() {
					personal_password := strings.SplitN(scanner.Text(), ":", 2)
					if len(personal_password) != 2 {
						log.Printf("Warning: Rejecting personal password setting on line %d (missing ':' delimiter)", line)
					} else {
						log.Printf("Set personal password for %s", personal_password[0])
						personalPasswords[personal_password[0]] = []byte(personal_password[1])
					}
				}
			}
		}
		fp.Close()
	}

	// start listening to incoming port
	incoming, err := net.Listen("tcp", fmt.Sprintf(":%d", *port))
	if err != nil {
		log.Fatalf("Unable to open incoming TCP port %d: %v", *port, err)
		os.Exit(2)
	}
	log.Printf("Listening on port %d", *port)
	defer incoming.Close()

	// signal handler
	sig_channel := make(chan os.Signal, 1)
	stop_channel := make(chan int, 1)
	signal.Notify(sig_channel, syscall.SIGHUP, syscall.SIGUSR1, syscall.SIGUSR2, syscall.SIGINT)

	ms := mapservice.MapService{
		IncomingListener:  incoming,
		Database:          sqldb,
		PlayerGroupPass:   groupPassword,
		GmPass:            gmPassword,
		PersonalPasswords: personalPasswords,
		Clients:           make(map[string]*mapservice.MapClient),
		InitFile:          *initfile,
		GreetFile:         *greetfile,
		EventHistory:      make(map[string]*mapservice.MapEvent),
		ImageList:         make(map[string]string),
		StopChannel:       stop_channel,
	}
	go ms.Run()
	go eventMonitor(sig_channel, stop_channel, &ms, *saveint)
	<-stop_channel
	log.Printf("Received STOP signal; shutting down")
	if err = ms.SaveState(); err != nil {
		log.Printf("Error trying to save game state before exit: %v", err)
	}
	ms.Shutdown()
	log.Printf("server shut down")
}

// @[00]@| GMA 4.2.2
// @[01]@|
// @[10]@| Copyright © 1992–2020 by Steven L. Willoughby
// @[11]@| (AKA MadScienceZone), Aloha, Oregon, USA. All Rights Reserved.
// @[12]@| Distributed under the terms and conditions of the BSD-3-Clause
// @[13]@| License as described in the accompanying LICENSE file distributed
// @[14]@| with GMA.
// @[15]@|
// @[20]@| Redistribution and use in source and binary forms, with or without
// @[21]@| modification, are permitted provided that the following conditions
// @[22]@| are met:
// @[23]@| 1. Redistributions of source code must retain the above copyright
// @[24]@|    notice, this list of conditions and the following disclaimer.
// @[25]@| 2. Redistributions in binary form must reproduce the above copy-
// @[26]@|    right notice, this list of conditions and the following dis-
// @[27]@|    claimer in the documentation and/or other materials provided
// @[28]@|    with the distribution.
// @[29]@| 3. Neither the name of the copyright holder nor the names of its
// @[30]@|    contributors may be used to endorse or promote products derived
// @[31]@|    from this software without specific prior written permission.
// @[32]@|
// @[33]@| THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
// @[34]@| CONTRIBUTORS “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES,
// @[35]@| INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// @[36]@| MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// @[37]@| DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
// @[38]@| BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
// @[39]@| OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// @[40]@| PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
// @[41]@| PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// @[42]@| THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
// @[43]@| TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
// @[44]@| THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
// @[45]@| SUCH DAMAGE.
// @[46]@|
// @[50]@| This software is not intended for any use or application in which
// @[51]@| the safety of lives or property would be at risk due to failure or
// @[52]@| defect of the software.
